File: 016. 3Sum Closest/solution-n^2.js
/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number}
 */

// Time: O(n^2)
var threeSumClosest = function(nums, target) {
    if (nums === null || nums.length < 3) return Number.MAX_SAFE_INTEGER;
    
    nums.sort(function(a, b) {
        return a - b;
    });
    
    var ret = Number.MAX_SAFE_INTEGER;
    for (var i = 0; i < nums.length - 2; i++) {
        var lo = i + 1;
        var hi = nums.length - 1;
        while (lo < hi) {
            var sum = nums[i] + nums[lo] + nums[hi];
            if (sum === target) return target;
            
            if (Math.abs(target - sum) < Math.abs(target - ret)) {
                ret = sum;
            }
            
            if (sum > target) {
                hi--;
            } else {
                lo++;
            }
        }
    }
    
    return ret;
};

File: Leet-016. 3Sum Closest/solution-n^2.java
class Solution {
    public int threeSumClosest(int[] nums, int target) {
        if (nums == null || nums.length < 3) {
            return Integer.MAX_VALUE;
        }
        
        int res = Integer.MAX_VALUE - 1000;
        Arrays.sort(nums);
        
        for (int i = 0; i < nums.length - 2; i++) {
            if (i > 0 && nums[i] == nums[i - 1]) {
                continue ;
            }
            
            int left = i + 1, right = nums.length - 1;
            while (left < right) {
                int sum = nums[i] + nums[left] + nums[right];
                
                if (Math.abs(sum - target) < Math.abs(res - target)) {
                    res = sum;
                }
                
                if (sum == target) {
                    return sum;
                } else if (sum < target) {
                    left++;
                } else {
                    right--;
                }
            }
        }
        
        return res;
    }
}